File: tests/mv-35/discs.js
#! /bin/env node
// we may need require to load re2
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
import * as runtime from "json_model_runtime"
const JSON_MODEL_VERSION = "2";

let _jm_map_0 = new Map()
export var check_model_map = new Map()

function _jm_obj_0(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        return false;
    }
    if (Object.keys(val).length != 2)
    {
        return false;
    }
    let pval;
    if (! val.hasOwnProperty("d"))
    {
        return false;
    }
    pval = val["d"];
    let res = ((typeof pval === 'string' || pval instanceof String)) && pval == "n";
    if (! res)
    {
        return false;
    }
    if (! val.hasOwnProperty("v"))
    {
        return false;
    }
    pval = val["v"];
    return pval === null;
}

function _jm_obj_1(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        return false;
    }
    if (Object.keys(val).length != 2)
    {
        return false;
    }
    let pval;
    if (! val.hasOwnProperty("d"))
    {
        return false;
    }
    pval = val["d"];
    let res = ((typeof pval === 'string' || pval instanceof String)) && pval == "b";
    if (! res)
    {
        return false;
    }
    if (! val.hasOwnProperty("v"))
    {
        return false;
    }
    pval = val["v"];
    return (typeof pval === 'boolean' || pval instanceof Boolean);
}

function _jm_obj_2(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        return false;
    }
    if (Object.keys(val).length != 2)
    {
        return false;
    }
    let pval;
    if (! val.hasOwnProperty("d"))
    {
        return false;
    }
    pval = val["d"];
    let res = ((typeof pval === 'string' || pval instanceof String)) && pval == "i";
    if (! res)
    {
        return false;
    }
    if (! val.hasOwnProperty("v"))
    {
        return false;
    }
    pval = val["v"];
    return (typeof pval === 'number' || pval instanceof Number) && Number.isInteger(pval);
}

function _jm_obj_3(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        return false;
    }
    if (Object.keys(val).length != 2)
    {
        return false;
    }
    let pval;
    if (! val.hasOwnProperty("d"))
    {
        return false;
    }
    pval = val["d"];
    let res = ((typeof pval === 'string' || pval instanceof String)) && pval == "u";
    if (! res)
    {
        return false;
    }
    if (! val.hasOwnProperty("v"))
    {
        return false;
    }
    pval = val["v"];
    return ((typeof pval === 'number' || pval instanceof Number) && Number.isInteger(pval)) && pval >= 0;
}

function _jm_obj_4(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        return false;
    }
    if (Object.keys(val).length != 2)
    {
        return false;
    }
    let pval;
    if (! val.hasOwnProperty("d"))
    {
        return false;
    }
    pval = val["d"];
    let res = ((typeof pval === 'string' || pval instanceof String)) && pval == "f";
    if (! res)
    {
        return false;
    }
    if (! val.hasOwnProperty("v"))
    {
        return false;
    }
    pval = val["v"];
    return (typeof pval === 'number' || pval instanceof Number);
}

function _jm_obj_5(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        return false;
    }
    if (Object.keys(val).length != 2)
    {
        return false;
    }
    let pval;
    if (! val.hasOwnProperty("d"))
    {
        return false;
    }
    pval = val["d"];
    let res = ((typeof pval === 'string' || pval instanceof String)) && pval == "s";
    if (! res)
    {
        return false;
    }
    if (! val.hasOwnProperty("v"))
    {
        return false;
    }
    pval = val["v"];
    return (typeof pval === 'string' || pval instanceof String);
}

function _jm_obj_6(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        return false;
    }
    if (Object.keys(val).length != 2)
    {
        return false;
    }
    let pval;
    if (! val.hasOwnProperty("d"))
    {
        return false;
    }
    pval = val["d"];
    let res = ((typeof pval === 'string' || pval instanceof String)) && pval == "a";
    if (! res)
    {
        return false;
    }
    if (! val.hasOwnProperty("v"))
    {
        return false;
    }
    pval = val["v"];
    return Array.isArray(pval);
}

function _jm_obj_7(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        return false;
    }
    if (Object.keys(val).length != 2)
    {
        return false;
    }
    let pval;
    if (! val.hasOwnProperty("d"))
    {
        return false;
    }
    pval = val["d"];
    let res = ((typeof pval === 'string' || pval instanceof String)) && pval == "o";
    if (! res)
    {
        return false;
    }
    if (! val.hasOwnProperty("v"))
    {
        return false;
    }
    pval = val["v"];
    return Object.prototype.toString.call(pval) === '[object Object]';
}


function json_model_1(val, path, rep)
{
    let res = Object.prototype.toString.call(val) === '[object Object]';
    if (res)
    {
        let tag_0;
        if ((tag_0 = val["d"]) != null)
        {
            let fun_0 = _jm_map_0.get(tag_0);
            res = fun_0 !== undefined && fun_0(val, null, null);
        }
        else
        {
            res = false;
        }
    }
    return res;
}


var initialized = false

// differed module initializations
export function check_model_init()
{
    if (! initialized)
    {
        initialized = true;
        runtime.jm_set_rx(RegExp)
        _jm_map_0.set("n", _jm_obj_0)
        _jm_map_0.set("b", _jm_obj_1)
        _jm_map_0.set("i", _jm_obj_2)
        _jm_map_0.set("u", _jm_obj_3)
        _jm_map_0.set("f", _jm_obj_4)
        _jm_map_0.set("s", _jm_obj_5)
        _jm_map_0.set("a", _jm_obj_6)
        _jm_map_0.set("o", _jm_obj_7)
        check_model_map.set("", json_model_1)
    }
}

// differed module cleanup
export function check_model_free()
{
    if (initialized)
    {
        initialized = false;
    }
}

export function check_model(val, name, rep)
{
    let checker = check_model_map.get(name)
    if (checker === undefined)
        throw `no checker for "${name}"`
    let path = rep !== null ? [] : null

    return checker(val, path, rep)
}

// possibly run as main based on a guess
import main from "json_model_runtime/main.js"

if (import.meta.url.endsWith(process.argv[1]))
    main(check_model_init, check_model, check_model_free)

File: tests/mv-35/discs.py
#! /bin/env python
import math
from typing import Callable
import re2 as re
from json_model.runtime import *
__version__ = "2"

def check_model_fun(name: str) -> CheckFun:
    """Return check function for JSON model name."""
    return check_model_map[name]

# entry point for generated checkers
def check_model(val: Jsonable, name: str = "", rep: Report = None) -> bool:
    """Check val validity against JSON Model name."""
    if name not in check_model_map:
        raise Exception(f"unexpected model name: {name}")
    checker = check_model_map[name]
    return checker(val, [], rep)

_jm_map_0: dict[str, str]
check_model_map: PropMap

def _jm_obj_0(val: Jsonable, path: Path, rep: Report) -> bool:
    if not isinstance(val, dict):
        return False
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("d", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, str) and pval == "n"
    if not res:
        return False
    if not ((pval := val.get("v", UNDEFINED)) != UNDEFINED):
        return False
    return pval is None

def _jm_obj_1(val: Jsonable, path: Path, rep: Report) -> bool:
    if not isinstance(val, dict):
        return False
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("d", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, str) and pval == "b"
    if not res:
        return False
    if not ((pval := val.get("v", UNDEFINED)) != UNDEFINED):
        return False
    return isinstance(pval, bool)

def _jm_obj_2(val: Jsonable, path: Path, rep: Report) -> bool:
    if not isinstance(val, dict):
        return False
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("d", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, str) and pval == "i"
    if not res:
        return False
    if not ((pval := val.get("v", UNDEFINED)) != UNDEFINED):
        return False
    return isinstance(pval, int) and not isinstance(pval, bool)

def _jm_obj_3(val: Jsonable, path: Path, rep: Report) -> bool:
    if not isinstance(val, dict):
        return False
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("d", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, str) and pval == "u"
    if not res:
        return False
    if not ((pval := val.get("v", UNDEFINED)) != UNDEFINED):
        return False
    return isinstance(pval, int) and not isinstance(pval, bool) and pval >= 0

def _jm_obj_4(val: Jsonable, path: Path, rep: Report) -> bool:
    if not isinstance(val, dict):
        return False
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("d", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, str) and pval == "f"
    if not res:
        return False
    if not ((pval := val.get("v", UNDEFINED)) != UNDEFINED):
        return False
    return isinstance(pval, float)

def _jm_obj_5(val: Jsonable, path: Path, rep: Report) -> bool:
    if not isinstance(val, dict):
        return False
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("d", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, str) and pval == "s"
    if not res:
        return False
    if not ((pval := val.get("v", UNDEFINED)) != UNDEFINED):
        return False
    return isinstance(pval, str)

def _jm_obj_6(val: Jsonable, path: Path, rep: Report) -> bool:
    if not isinstance(val, dict):
        return False
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("d", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, str) and pval == "a"
    if not res:
        return False
    if not ((pval := val.get("v", UNDEFINED)) != UNDEFINED):
        return False
    return isinstance(pval, list)

def _jm_obj_7(val: Jsonable, path: Path, rep: Report) -> bool:
    if not isinstance(val, dict):
        return False
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("d", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, str) and pval == "o"
    if not res:
        return False
    if not ((pval := val.get("v", UNDEFINED)) != UNDEFINED):
        return False
    return isinstance(pval, dict)


def json_model_1(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool = isinstance(val, dict)
    if res:
        tag_0: Jsonable
        if (tag_0 := val.get("d", UNDEFINED)) != UNDEFINED:
            fun_0: CheckFun = _jm_map_0.get(tag_0, UNDEFINED) if not isinstance(tag_0, (list, dict)) else UNDEFINED
            res = fun_0 != UNDEFINED and fun_0(val, None, None)
        else:
            res = False
    return res


# initialization guard
initialized: bool = False

# differed module initializations
def check_model_init():
    global initialized
    if not initialized:
        initialized = True
        global _jm_map_0
        _jm_map_0 = {
            "n": _jm_obj_0,
            "b": _jm_obj_1,
            "i": _jm_obj_2,
            "u": _jm_obj_3,
            "f": _jm_obj_4,
            "s": _jm_obj_5,
            "a": _jm_obj_6,
            "o": _jm_obj_7,
        }
        global check_model_map
        check_model_map = {
            "": json_model_1,
        }

# differed module cleanup
def check_model_free():
    global initialized
    if initialized:
        initialized = False

if __name__ == "__main__":
    check_model_init()
    main(check_model_fun, check_model_map, __version__)
    check_model_free()
